test/config/normalizers.test.js

```diff
@@ -0,0 +1,388 @@
+
+  ]
+  const defaults = { badWithDefault: '25s' }
+  const opts = {
+    withoutUnit: '200',
+    withUnit: '2m',
+    notAllowedUnit: '20us',
+    notAllowedNegative: '-1s',
+    badWithDefault: 'not-duration',
+    badWithoutDefault: 'not-duration',
+    anotherProperty: 25
+  }
+
+  normalizeDurationOptions(opts, fields, defaults, logger)
+
+  t.deepEqual(opts, {
+    // TODO: this normalized deletes keys whose values are not valid (because format or contraint)
+    // other normalizers keep the property as is. Should we remove as well?
+    withoutUnit: 0.2,
+    withUnit: 120,
+    badWithDefault: 25,
+    anotherProperty: 25
+  })
+
+  const warnings = logger.calls
+  t.ok(warnings.length === 4, 'we got warnings for bad duration options')
+  t.ok(warnings[0].message.indexOf('ignoring this option') !== -1, 'ignores not allowed unit')
+  t.deepEqual(warnings[0].interpolation, ['20us', 'notAllowedUnit'])
+  t.ok(warnings[1].message.indexOf('ignoring this option') !== -1, 'ignores not allowed negative value')
+  t.deepEqual(warnings[1].interpolation, ['-1s', 'notAllowedNegative'])
+  t.ok(warnings[2].message.indexOf('using default') !== -1, 'uses default value')
+  t.deepEqual(warnings[2].interpolation, ['not-duration', 'badWithDefault', '25s'])
+  t.ok(warnings[3].message.indexOf('ignoring this option') !== -1, 'ignores bad value without default')
+  t.deepEqual(warnings[3].interpolation, ['not-duration', 'badWithoutDefault'])
+  t.end()
+})
+
+test('#normalizeIgnoreOptions()', function (t) {
+  const logger = new MockLogger()
+  const defaults = {}
+  const opts = {
+    transactionIgnoreUrls: ['*path*'],
+    ignoreUrls: ['str/path/ignore', /path\/to\/secret/],
+    ignoreUserAgents: ['Mozilla', /Safari/],
+    ignoreMessageQueues: ['*topic*']
+  }
+
+  normalizeIgnoreOptions(opts, Object.keys(opts), defaults, logger)
+
+  t.deepEqual(opts, {
+    transactionIgnoreUrls: ['*path*'],
+    transactionIgnoreUrlRegExp: [/^.*path.*$/i],
+    ignoreUrlStr: ['str/path/ignore'],
+    ignoreUrlRegExp: [/path\/to\/secret/],
+    ignoreUserAgentStr: ['Mozilla'],
+    ignoreUserAgentRegExp: [/Safari/],
+    ignoreMessageQueues: ['*topic*'],
+    ignoreMessageQueuesRegExp: [/^.*topic.*$/i]
+  })
+
+  t.ok(logger.calls.length === 0, 'we got no warnings for bad ignore options')
+  t.end()
+})
+
+test('#normalizeInfinity()', function (t) {
+  const logger = new MockLogger()
+  const fields = ['minusOne']
+  const defaults = {}
+  const opts = {
+    positiveNumber: 100,
+    negativeNumber: -100,
+    minusOne: -1,
+    anotherProperty: 'value'
+  }
+
+  normalizeInfinity(opts, fields, defaults, logger)
+
+  t.deepEqual(opts, {
+    positiveNumber: 100,
+    negativeNumber: -100,
+    minusOne: Infinity,
+    anotherProperty: 'value'
+  })
+
+  t.ok(logger.calls.length === 0, 'we got no warnings for bad ignore options')
+  t.end()
+})
+
+test.skip('#normalizeKeyValuePairs()', function (t) {
```

Comment:
this test is skipped because either there is a bug or I did not get the spec properly. See comment in the test corpus

Reply:
Took me a while to see it. The `return` in `normalizeKeyValuePairs()` is a bug. It means that processing stops after the first key that is an object -- in this case only `objectProperty` is handled.  This was broken back in https://github.com/elastic/apm-agent-nodejs/pull/1027    Nice job finding issues with this refactoring. :)

It means that any user that was attempting to use both `globalLabels` and `addPatch` at the same time might hit the bug. I would be shocked if any users are using `addPatch` -- it is really low level.

I think you might want to *not* fix this, but wait for a separate change to fix it. Then this change can remain a "no functional change" PR.

The `return` should be `continue`, I think. Or better yet, this implementation should be rewritten to use a switch and not be so subtle in which branches are intended to be taken. It also blows up if a `null` value is given (because `typeof null === 'object'`):

With this change:

```diff
diff --git a/test/config/normalizers.test.js b/test/config/normalizers.test.js
index 275c0d6a..90155e2a 100644
--- a/test/config/normalizers.test.js
+++ b/test/config/normalizers.test.js
@@ -198,10 +198,11 @@ test('#normalizeInfinity()', function (t) {
   t.end()
 })

-test.skip('#normalizeKeyValuePairs()', function (t) {
+test.only('#normalizeKeyValuePairs()', function (t) {
   const logger = new MockLogger()
   const defaults = {}
   const opts = {
+    badNull: null,
     objectProperty: { foo: 'bar', eggs: 'spam' },
     stringProperty: 'foo=bar, eggs=spam',
     arrayProperty: ['foo=bar', 'eggs=spam'],
```

I get this crash:

```
% node test/config/normalizers.test.js
TAP version 13
# #normalizeKeyValuePairs()
/Users/trentm/el/apm-agent-nodejs12/lib/config/normalizers.js:39
        opts[key] = Object.entries(opts[key])
                           ^

TypeError: Cannot convert undefined or null to object
    at Function.entries (<anonymous>)
    at normalizeKeyValuePairs (/Users/trentm/el/apm-agent-nodejs12/lib/config/normalizers.js:39:28)
    at Test.<anonymous> (/Users/trentm/el/apm-agent-nodejs12/test/config/normalizers.test.js:213:3)
    at Test.bound [as _cb] (/Users/trentm/el/apm-agent-nodejs12/node_modules/tape/lib/test.js:95:17)
    at Test.run (/Users/trentm/el/apm-agent-nodejs12/node_modules/tape/lib/test.js:115:28)
    at Test.bound [as run] (/Users/trentm/el/apm-agent-nodejs12/node_modules/tape/lib/test.js:95:17)
    at Immediate.next (/Users/trentm/el/apm-agent-nodejs12/node_modules/tape/lib/results.js:157:7)
    at processImmediate (node:internal/timers:466:21)
```